File: libtiledbsoma/test/test_indexer.py
import numpy as np
import pandas as pd

from tiledbsoma.IntIndexer import IntIndexer

# tiledbsoma.pytiledbsoma.config_logging("debug")


# 1d array to list


def indexer_test(keys: np.array, lookups: np.array, fail: bool):
    if fail:
        indexer_test_fail(keys, lookups)
    else:
        indexer_test_pass(keys, lookups)


def indexer_test_fail(keys: np.array, lookups: np.array):
    reindexing_exception: bool = False
    try:
        indexer = IntIndexer.map_locations(keys)
        indexer.get_indexer(lookups)
    except pd.errors.InvalidIndexError:
        reindexing_exception = True

    try:
        panda_indexer = pd.Index(keys)
        panda_indexer.get_indexer(lookups)
    except pd.errors.InvalidIndexError as exception:
        if reindexing_exception:
            return
        raise exception


def indexer_test_pass(keys: np.array, lookups: np.array):
    indexer = IntIndexer.map_locations(keys)
    results = indexer.get_indexer(lookups)
    panda_indexer = pd.Index(keys)
    panda_results = panda_indexer.get_indexer(lookups)
    assert np.equal(results.all(), panda_results.all())


test_data = [
    {
        "keys": [-1, -1, -1, 0, 0, 0],
        "lookups": [1, 2, 3, 4, 5, 1, 2, 3, 4, 5, 1, 2, 3, 4, 5, 1, 2, 3, 4, 5],
        "pass": False,
    },
    {
        "keys": [
            -1,
            1,
            2,
            3,
            4,
            5,
            -1,
            1,
            2,
            3,
            4,
            5,
            -1,
            1,
            2,
            3,
            4,
            5,
            -1,
            1,
            2,
            3,
            4,
            5,
        ],
        "lookups": [-10000, 1, 2, 3, 5, 6],
        "pass": False,
    },
    {
        "keys": [-1, 1, 2, 3, 4, 5],
        "lookups": [
            -1,
            1,
            2,
            3,
            4,
            5,
            -1,
            1,
            2,
            3,
            4,
            5,
            -1,
            1,
            2,
            3,
            4,
            5,
            -1,
            1,
            2,
            3,
            4,
            5,
        ],
        "pass": True,
    },
    {
        "keys": [-10000, -100000, 200000, 5, 1, 7],
        "lookups": [
            -1,
            1,
            2,
            3,
            4,
            5,
            -1,
            1,
            2,
            3,
            4,
            5,
            -1,
            1,
            2,
            3,
            4,
            5,
            -1,
            1,
            2,
            3,
            4,
            5,
        ],
        "pass": True,
    },
    {
        "keys": [-10000, -200000, 1000, 3000, 1, 2],
        "lookups": [
            -1,
            1,
            2,
            3,
            4,
            5,
            -1,
            1,
            2,
            3,
            4,
            5,
            -1,
            1,
            2,
            3,
            4,
            5,
            -1,
            1,
            2,
            3,
            4,
            5,
        ],
        "pass": True,
    },
    {
        "keys": [i for i in range(1, 10000)],
        "lookups": [i for i in range(1, 10)],
        "pass": True,
    },
    {
        "keys": [1, 2, 3, 4, 5, 2**63 - 1],
        "lookups": [i for i in range(2**63 - 1000, 2**63 - 1)],
        "pass": True,
    },
]


def main():
    for data in test_data:
        indexer_test(data["keys"], data["lookups"], not data["pass"])


if __name__ == "__main__":
    main()

File: libtiledbsoma/test/test_indexer.cc
/**
 * @file   test_indexer.cc
 *
 * @section LICENSE
 *
 * Licensed under the MIT License.
 * Copyright (c) TileDB, Inc. and The Chan Zuckerberg Initiative Foundation
 *
 * @section DESCRIPTION
 *
 * This file manages unit tests for re-indexer
 */

#include <reindexer/reindexer.h>
#include <catch2/catch_test_macros.hpp>
#include <cstdint>
#include <string>
#include <tiledb/tiledb>
#include <unordered_map>
#include <vector>
#include "external/khash/khash.h"

namespace {

/**
 * @brief Create multiple 64 bit vector and try the reindexer compared to khash
 * as baseline.
 *
 */

KHASH_MAP_INIT_INT64(m64, int64_t)

// Identify the test in which keys are not unique and we should expect
// exception.
std::vector<bool> uniqueness = {false, false, true, true, true};

bool run_test(size_t id, std::vector<int64_t> keys, std::vector<int64_t> lookups) {
    try {
        std::vector<int64_t> indexer_results;
        indexer_results.resize(lookups.size());

        tiledbsoma::IntIndexer indexer;
        indexer.map_locations(keys);
        auto* hash = kh_init(m64);
        int ret;
        khint64_t k;

        for (size_t i = 0; i < keys.size(); i++) {
            k = kh_put(m64, hash, keys[i], &ret);
            assert(k != kh_end(hash));
            kh_val(hash, k) = i;
        }

        indexer.lookup(lookups, indexer_results);
        std::vector<int64_t> kh_results;
        kh_results.resize(lookups.size());
        for (size_t i = 0; i < lookups.size(); i++) {
            auto k = kh_get(m64, hash, lookups[i]);
            if (k == kh_end(hash)) {
                // According to pandas behavior
                kh_results[i] = -1;
            } else {
                kh_results[i] = kh_val(hash, k);
            }
        }
        kh_destroy(m64, hash);
        return indexer_results == kh_results;
    } catch (const std::runtime_error& e) {
        // Path with non unique keys
        if (!uniqueness[id]) {
            return true;
        }
        return false;
    }
}

// Test data
std::vector<std::unordered_map<std::string, std::vector<int64_t>>> test_data = {
    {
        {"keys", {-1, -1, -1, 0, 0, 0}},
        {"lookups", {1, 2, 3, 4, 5, 1, 2, 3, 4, 5, 1, 2, 3, 4, 5, 1, 2, 3, 4, 5}},
    },
    {{"keys",
      {
          -1, 1, 2, 3, 4, 5, -1, 1, 2, 3, 4, 5, -1, 1, 2, 3, 4, 5, -1, 1, 2, 3, 4, 5,
      }},
     {"lookups", {-10000, 1, 2, 3, 5, 6}}},
    {
        {"keys", {-1, 1, 2, 3, 4, 5}},
        {"lookups",
         {
             -1, 1, 2, 3, 4, 5, -1, 1, 2, 3, 4, 5, -1, 1, 2, 3, 4, 5, -1, 1, 2, 3, 4, 5,
         }},
    },
    {
        {"keys", {-10000, -100000, 200000, 5, 1, 7}},
        {"lookups",
         {
             -1, 1, 2, 3, 4, 5, -1, 1, 2, 3, 4, 5, -1, 1, 2, 3, 4, 5, -1, 1, 2, 3, 4, 5,
         }},
    },
    {
        {"keys", {-10000, -200000, 1000, 3000, 1, 2}},
        {"lookups",
         {
             -1, 1, 2, 3, 4, 5, -1, 1, 2, 3, 4, 5, -1, 1, 2, 3, 4, 5, -1, 1, 2, 3, 4, 5,
         }},
    }};

TEST_CASE("C++ re-indexer") {
    for (size_t test = 0; test < test_data.size(); test++) {
        bool result = run_test(test, test_data[test]["keys"], test_data[test]["lookups"]);
        if (!result) {
            throw std::runtime_error("Test " + std::to_string(test) + " failed");
        }
    }
}
}  // namespace
